lib/SILOptimizer/Mandatory/PredictableMemOpt.cpp

```diff
@@ -2516,6 +2628,144 @@ bool AllocOptimize::tryToRemoveDeadAllocation() {
+  // is known dynamically none along a path. This is dynamically correct, but
+  // can not be represented in OSSA so we insert these destroys along said path.
+  SmallVector<SILBasicBlock *, 32> consumingUseBlocks;
+  while (!valuesNeedingLifetimeCompletion.empty()) {
+    auto optV = valuesNeedingLifetimeCompletion.pop_back_val();
+    if (!optV)
+      continue;
+    SILValue v = *optV;
+    if (v.getOwnershipKind() != OwnershipKind::Owned)
+      continue;
+
+    // First see if our value doesn't have any uses. In such a case, just
+    // insert a destroy_value at the next instruction and return.
+    if (v->use_empty()) {
+      auto *next = v->getNextInstruction();
+      auto loc = RegularLocation::getAutoGeneratedLocation();
+      SILBuilderWithScope localBuilder(next);
+      localBuilder.createDestroyValue(loc, v);
+      continue;
+    }
+
+    // Otherwise, we first see if we have any consuming uses at all. If we do,
+    // then we know that any such consuming uses since we have an owned value
+    // /must/ be strongly control equivalent to our value and unreachable from
+    // each other, so we can just use findJointPostDominatingSet to complete
+    // the set.
+    consumingUseBlocks.clear();
+    for (auto *use : v->getConsumingUses())
+      consumingUseBlocks.push_back(use->getParentBlock());
+
+    if (!consumingUseBlocks.empty()) {
+      findJointPostDominatingSet(
+          v->getParentBlock(), consumingUseBlocks, [](SILBasicBlock *) {},
+          [&](SILBasicBlock *result) {
+            auto loc = RegularLocation::getAutoGeneratedLocation();
+            SILBuilderWithScope builder(result);
+            builder.createDestroyValue(loc, v);
+          });
+      continue;
+    }
+
+    // If we do not have at least one consuming use, we need to do something
+    // different. This situation can occur given a non-trivial enum typed
+    // stack allocation that:
+    //
+    // 1. Had a destroy_addr eliminated along a path where we dynamically know
+    //    that the stack allocation is storing a trivial case.
+    //
+    // 2. Had some other paths where due to dead end blocks, no destroy_addr
+    //    is needed.
+    //
+    // To fix this, we just treat all uses as consuming blocks and insert
+    // destroys using the joint post dominance set computer and insert
+    // destroys at the end of all input blocks in the post dom set and at the
+    // beginning of any leaking blocks.
+    {
+      // TODO: Can we just pass this in to findJointPostDominatingSet instead
+      // of recomputing it there? Maybe an overload that lets us do this?
+      BasicBlockSet foundUseBlocks(v->getFunction());
+      for (auto *use : v->getUses()) {
+        auto *block = use->getParentBlock();
+        if (!foundUseBlocks.insert(block))
+          continue;
+        consumingUseBlocks.push_back(block);
+      }
+    }
+    findJointPostDominatingSet(
```

Comment:
fyi, the whole point of ValueLifetimeAnalysis is to do what you're doing manually here. That utility has been corrupted and is broken, so you shouldn't use it as-is, but I have a fixed, pure version on a branch that just does what you're doing manually and I'll put it up ASAP. Here you're putting destroys in successor blocks when they don't need to be, which seems like a bad idea but maybe was expedient.

Reply:
SGTM